### DISCUSSION TITLE
How do I build a 'brand new' release of OKD that is not based on a prior release

### DISCUSSION THREAD

**Discussion by fortinj66** [okd-project/okd#1231] [General]
We've talked multiple times about building OKD at a component level (ie rebuild MCO, or console) and replacing those components in an already existing build...

Something like this:

```
[johnf@johnf-dev ~]$ cat make_payload.sh
server=https://api.ci.openshift.org

from_release=quay.io/openshift/okd:4.10.0-0.okd-2022-04-14-012059
to_image=quay.io/fortinj66/origin-release:v4.10-2022-04-19-pin-kernel
release_name=4.10.0-0.jef-2022-04-19-pin-kernel

oc adm release new --from-release ${from_release} \
                   --name ${release_name} \
                   --to-image ${to_image} \
                   machine-os-content=quay.io/fortinj66/okd-machine-os:4.10-pinned-kernel
```

That's great and works great for testing fixes, but there are some issues.  

For example, if I look at the resulting image:

```
[johnf@johnf-dev ~]$ oc adm release info quay.io/fortinj66/origin-release:v4.10-2022-04-19-pin-kernel
Name:      4.10.0-0.jef-2022-04-19-pin-kernel
Digest:    sha256:0401d82dab5f324fa89ca0fe8331659d5724e72ab37e72bd9c384d822f9fff86
Created:   2022-04-19T19:36:34Z
OS/Arch:   linux/amd64
Manifests: 571

Pull From: quay.io/fortinj66/origin-release@sha256:0401d82dab5f324fa89ca0fe8331659d5724e72ab37e72bd9c384d822f9fff86

Release Metadata:
  Version:  4.10.0-0.jef-2022-04-19-pin-kernel
  Upgrades: <none>
  Metadata:

Component Versions:
  kubernetes 1.23.5
  machine-os 410.35.3 Fedora CoreOS

Images:
  NAME                                           DIGEST
  alibaba-cloud-controller-manager               sha256:d61708b0a9f9a15b70011630573e59ae8eb242393c9aeebcec4c0d925e170823
  alibaba-cloud-csi-driver                       sha256:ac0cd7e863216978046247f66ca6e677d9af302178f7759245b1fc1247b1594f
  alibaba-disk-csi-driver-operator               sha256:27375601f535e3d9a907fe7eb7a64a623935b00e0f630e1cdff794cf78447907
...
  machine-api-operator                           sha256:a550b82a351df9aea2cd3a6ed4d5a8b20201d4d9df4d2d35ec960ea4f99fa90d
  machine-config-operator                        sha256:f904b4903dfc7c574fc2bc32b29c8c24d57097c4445cc16095f8c22c4ea6394d
  machine-os-content                             quay.io/fortinj66/okd-machine-os:4.10-pinned-kernel
  machine-os-images                              sha256:e1c32da05f34adc0fbd51f43798662490c718262a92a8b4e2339da5a6f56e585
...
```

you can see I do not get a shaxxx for `machine-os-content .`  I get 

`  machine-os-content                             quay.io/fortinj66/okd-machine-os:4.10-pinned-kernel`

Nor do I get a commit-url for `machine-os-content`
```
[johnf@johnf-dev ~]$ oc adm release info quay.io/fortinj66/origin-release:v4.10-2022-04-19-pin-kernel --commit-urls
Name:      4.10.0-0.jef-2022-04-19-pin-kernel
Digest:    sha256:0401d82dab5f324fa89ca0fe8331659d5724e72ab37e72bd9c384d822f9fff86
Created:   2022-04-19T19:36:34Z
OS/Arch:   linux/amd64
Manifests: 571

Pull From: quay.io/fortinj66/origin-release@sha256:0401d82dab5f324fa89ca0fe8331659d5724e72ab37e72bd9c384d822f9fff86

Release Metadata:
  Version:  4.10.0-0.jef-2022-04-19-pin-kernel
  Upgrades: <none>
  Metadata:

Component Versions:
  kubernetes 1.23.5
  machine-os 410.35.3 Fedora CoreOS

Images:
  NAME                                           URL
...
  machine-config-operator                        https://github.com/openshift/machine-config-operator/commit/14a1ca2cb91ff7e0faf9146b21ba12cd6c652d22
  machine-os-content
  machine-os-images                              https://github.com/openshift/machine-os-images/commit/4d1e087ed00fddbab16a8cf8fe22d0cce4fef025
...  
```

So, how do we create a new release image that has all this information, including from custom images, but does not depend on a currently existing image for the metadata.  

I realize that a lot of this magic happens in @vrutkovs backend release process, but I would like to understand it and maybe reproduce it myself.

**Comment by vrutkovs** [comment]
>you can see I do not get a shaxxx for machine-os-content 

Yeah, `oc adm release new` can't resolve images when a new release is created (and most likely should not). `machine-os-content=quay.io/fortinj66/okd-machine-os@sha256:foobar` should work.


>Nor do I get a commit-url for machine-os-content

Commit URL is retrieved from image labels:
```
$ oc adm release info quay.io/openshift/okd:4.10.0-0.okd-2022-03-07-131213 --commit-urls | grep installer                                     
  baremetal-installer                            https://github.com/openshift/installer/commit/3b701903d96b6375f6c3852a02b4b70fea01d694
  installer                                      https://github.com/openshift/installer/commit/3b701903d96b6375f6c3852a02b4b70fea01d694
  installer-artifacts                            https://github.com/openshift/installer/commit/3b701903d96b6375f6c3852a02b4b70fea01d694
$ oc adm release info quay.io/openshift/okd:4.10.0-0.okd-2022-03-07-131213 --image-for=installer                                        
quay.io/openshift/okd-content@sha256:fe79d5cc1dc101e4d8bcf77ea780559ed0fa095060026eeb50cd9f77aabf6f32
$ skopeo inspect -n docker://quay.io/openshift/okd-content@sha256:fe79d5cc1dc101e4d8bcf77ea780559ed0fa095060026eeb50cd9f77aabf6f32 | grep io.openshift.build.commit.id
        "io.openshift.build.commit.id": "3b701903d96b6375f6c3852a02b4b70fea01d694",
```

As for "entirely new release from scratch" - I'm not sure how to implement this. CI continuously builds new releases based on the existing ones, starting from early 4.0 days

**Reply by fortinj66** [reply]
> > you can see I do not get a shaxxx for machine-os-content
> 
> Yeah, `oc adm release new` can't resolve images when a new release is created (and most likely should not). `machine-os-content=quay.io/fortinj66/okd-machine-os@sha256:foobar` should work.
> 
> > Nor do I get a commit-url for machine-os-content
> 
> Commit URL is retrieved from image labels:
> 
> ```
> $ oc adm release info quay.io/openshift/okd:4.10.0-0.okd-2022-03-07-131213 --commit-urls | grep installer                                     
>   baremetal-installer                            https://github.com/openshift/installer/commit/3b701903d96b6375f6c3852a02b4b70fea01d694
>   installer                                      https://github.com/openshift/installer/commit/3b701903d96b6375f6c3852a02b4b70fea01d694
>   installer-artifacts                            https://github.com/openshift/installer/commit/3b701903d96b6375f6c3852a02b4b70fea01d694
> $ oc adm release info quay.io/openshift/okd:4.10.0-0.okd-2022-03-07-131213 --image-for=installer                                        
> quay.io/openshift/okd-content@sha256:fe79d5cc1dc101e4d8bcf77ea780559ed0fa095060026eeb50cd9f77aabf6f32
> $ skopeo inspect -n docker://quay.io/openshift/okd-content@sha256:fe79d5cc1dc101e4d8bcf77ea780559ed0fa095060026eeb50cd9f77aabf6f32 | grep io.openshift.build.commit.id
>         "io.openshift.build.commit.id": "3b701903d96b6375f6c3852a02b4b70fea01d694",
> ```
Hmmm, I wonder when that is being added to the images...  I don't recall seeing it in the build process itself, at least for okd-machine-os

> 
> As for "entirely new release from scratch" - I'm not sure how to implement this. CI continuously builds new releases based on the existing ones, starting from early 4.0 days

This is sort of a chicken and egg syndrome...  How do we get a list of current images and package them into a release, but you need a release to get the list of images...

**Reply by vrutkovs** [reply]
>Hmmm, I wonder when that is being added to the images... I don't recall seeing it in the build process itself, at least for okd-machine-os

This is done by openshift's BuildConfig controller automatically (provided buildconfig source is git)

**Comment by bdlink** [comment]
Is this the same issue as "how do I build OKD from source, including all components?"  

I think the many people who have attempted this have had numerous questions and difficulties, some of which are below (and my questions may well be simplistic, too - my apologies in advance).  Props to John for opening this discussion thread.  

For example, how does one know what the reference source is for each component (manifests, images).  For example, the first image is  alibaba-cloud-controller-manager.  Where does that come from, specifically, wrt source?

Is an OKD-specific version just a layer on top of an OCP or K8S container, or does it depend on the image?

It sounds like running Prow ourselves will not help (in terms of building from source).

As I understand the principles of source code control, building release n+1 on top of release n is not enough to have reproducibility.  You have to be able to go the the source.  Assuming OCP is done the same way, this process creates some hazards (including not being able to read the source behind the image looking for security issues)

As well, of course, there are Dockerfiles where the "FROM" image is unavailable, e.g. registry.ci.openshift.org/ocp/builder:rhel-8-golang-1.17-openshift-4.11 for the installer.  How many images are built on proprietary base images, and presumably are themselves proprietary?  The installer repo I was looking at is an OCP one, but linked from an OKD page.  

Maybe we need a list of issues/roadblocks on the way to being able to build from src with everything being open-source.  

With Kubernetes, there is easy documentation on building it from source, as well as running tests on the results, see https://github.com/kubernetes/kubernetes/blob/master/README.md This would be a good goal for an open-source OKD.

**Reply by vrutkovs** [reply]
>For example, the first image is alibaba-cloud-controller-manager. Where does that come from, specifically, wrt source?

```
$ oc adm release info quay.io/openshift/okd:4.10.0-0.okd-2022-03-07-131213 --commit-urls | grep alibaba                            
  alibaba-cloud-controller-manager               https://github.com/openshift/cloud-provider-alibaba-cloud/commit/db2d118ad70ff62a2111e83a8d14c5b32e176b38
  alibaba-cloud-csi-driver                       https://github.com/openshift/alibaba-cloud-csi-driver/commit/b9dd1ad5bd2a8971c22dca927d61d8f259516035
  alibaba-disk-csi-driver-operator               https://github.com/openshift/alibaba-disk-csi-driver-operator/commit/9c440d2b64cd30d2e56f3d1ccee1147151540ccc
  alibaba-machine-controllers                    https://github.com/openshift/cluster-api-provider-alibaba/commit/0206121348c9a0d220dd6805cea79d1eae7fd3e0
```
See https://github.com/openshift/okd/blob/master/CONTRIBUTING.md#replacing-other-images

>Is an OKD-specific version just a layer on top of an OCP or K8S container, or does it depend on the image?

Unless its in [the exception list](https://github.com/openshift/release/blob/master/core-services/release-controller/_releases/release-ocp-4.10-ci.json#L11-L42) its being copied verbatim from OCP CI. All exception images have their own custom configuration i.e. [okd-machine-os](https://github.com/openshift/release/blob/master/ci-operator/config/openshift/okd-machine-os/openshift-okd-machine-os-release-4.10.yaml#L21-L25) or [okd installer](https://github.com/openshift/release/blob/master/ci-operator/config/openshift/installer/openshift-installer-release-4.10__okd.yaml#L101-L103)

> It sounds like running Prow ourselves will not help (in terms of building from source).

Right, Prow is just a CI coordinator, all build rules are defined in https://github.com/openshift/release, regardless of how the images get built

>As well, of course, there are Dockerfiles where the "FROM" image is unavailable, e.g. registry.ci.openshift.org/ocp/builder:rhel-8-golang-1.17-openshift-4.11 for the installer

If we make sure the base image doesn't contain RHEL packages we can make it public and redistributable. Most likely its not possible :/

>Maybe we need a list of issues/roadblocks on the way to being able to build from src with everything being open-source.

Strictly speaking we don't need to make exact same builder images public. While we do include Dockerfile for our CI, its not necessarily the only way to have the source code built as a container image. 

However it would certainly useful for the community. As an experiment we could try replacing `rhel-8-golang-1.17-openshift-4.11` images with "Centos8 Stream base with golang 1.17 installed" and see if all (or at least most) images can be rebuilt. 
If that works we can add OKD-specific instruction on how to replace the base image for most images - and then work on making builder images public.

**Reply by fortinj66** [reply]
> > As well, of course, there are Dockerfiles where the "FROM" image is unavailable, e.g. registry.ci.openshift.org/ocp/builder:rhel-8-golang-1.17-openshift-4.11 for the installer
> 
> If we make sure the base image doesn't contain RHEL packages we can make it public and redistributable. Most likely its not possible :/
> 
> > Maybe we need a list of issues/roadblocks on the way to being able to build from src with everything being open-source.
> 
> Strictly speaking we don't need to make exact same builder images public. While we do include Dockerfile for our CI, its not necessarily the only way to have the source code built as a container image.
> 
> However it would certainly useful for the community. As an experiment we could try replacing `rhel-8-golang-1.17-openshift-4.11` images with "Centos8 Stream base with golang 1.17 installed" and see if all (or at least most) images can be rebuilt. If that works we can add OKD-specific instruction on how to replace the base image for most images - and then work on making builder images public.

@LorbusChris was going to look at possibly having those Dockerfiles reconfigured with an available image as that image gets replaced during the CI build process anyway.  I'm not sure where he is with that.

**Comment by bdlink** [comment]
"oc adm release info quay.io/openshift/okd:4.10.0-0.okd-2022-03-07-131213 --commit-urls" 

Thanks, I had forgotten that.  

So, for all of the images, as well as the machine os content, the source (if not the base images, as well as FCOS) is under github.com/openshift.  This includes a fork (?) of Kubernetes under github.com/openshift/kubernetes
